### DISCUSSION TITLE
Embedding Xonsh

### DISCUSSION THREAD

**Discussion by jamie-at-arkifi** [xonsh/xonsh#5260] [Q&A] [answered]
What's the advice for embedding xonsh these days? That is, I have a Python project, and I would like to write some of it in Xonsh instead of Python.

EDIT: I'm currently doing this:

```
import xonsh.imphooks
xonsh.imphooks.install_import_hooks(None)  # noqa
```

but I don't know if that's current best practice and will correctly handle things like environment isolation.

**Comment by anki-code** [comment]
Start from the [cheatsheet](https://github.com/anki-code/xonsh-cheatsheet).

**Reply by jamie-at-arkifi** [reply]
Wrong kind of embedding.

I'm currently using

```
import xonsh.imphooks
xonsh.imphooks.install_import_hooks(None)  # noqa
```

But it's been a while since I dug into Xonsh this deeply, so I wanted to check.

**Comment by jamie-at-arkifi** [comment, marked as answer]
Here is the full and correct embedding, I think:

```xsh
import xonsh.built_ins
import xonsh.execer
import xonsh.imphooks

execer = xonsh.execer.Execer()
xonsh.built_ins.XSH.load(execer=execer)
xonsh.imphooks.install_import_hooks(execer=execer) 
```

Given the level of init going on, I'm guessing this doesn't read xonshrc or similar. But that's just conjecture.

I might hop on my personal account and submit this as a PR.

**Reply by anki-code** [reply]
Thanks @jamie-at-arkifi ! It will be cool to have real life example of using this because the "embedding" word is very common and personally I don't clearly understand what you mean. And PR is welcome!

**Reply by jamie-at-arkifi** [reply]
I'm using it in my pulumi code, so i can write xsh modules.
